From: Pavel Hrdina <phrdina@redhat.com>
Date: Thu, 22 Sep 2016 19:37:47 +0200
Subject: [PATCH] console: add support to forget password

If password for console is saved currently there is no way how to tell
virt-manager to forget that password.  This patch improves the authentication
page in order to provide a way how to forget password simply by unchecking the
"Save this password in your keyring".

Signed-off-by: Pavel Hrdina <phrdina@redhat.com>
(cherry picked from commit 09cc6f3832ec03d54e8d98776aa1bcb51f428b52)

Resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1301841

Signed-off-by: Pavel Hrdina <phrdina@redhat.com>
---
 src/virtManager/config.py  | 13 +++++++++++++
 src/virtManager/console.py |  2 ++
 src/virtManager/keyring.py |  6 ++++++
 src/vmm-details.glade      |  1 +
 4 files changed, 22 insertions(+)

diff --git a/src/virtManager/config.py b/src/virtManager/config.py
index 5e3af1b..e14e2ce 100644
--- a/src/virtManager/config.py
+++ b/src/virtManager/config.py
@@ -831,3 +831,16 @@ class vmmConfig(object):
         if _id != None:
             self.conf.set_int(self.conf_dir + "/console/passwords/" + vm.get_uuid(), _id)
             self.conf.set_string(self.conf_dir + "/console/usernames/" + vm.get_uuid(), username)
+
+    def del_console_password(self, vm):
+        if not self.has_keyring():
+            return
+
+        keyid = self.conf.get_int(self.conf_dir + "/console/passwords/" + vm.get_uuid())
+
+        if keyid == -1:
+            return
+
+        self.keyring.del_secret(keyid)
+        self.conf.set_int(self.conf_dir + "/console/passwords/" + vm.get_uuid(), -1)
+        self.conf.set_string(self.conf_dir + "/console/usernames/" + vm.get_uuid(), "")
diff --git a/src/virtManager/console.py b/src/virtManager/console.py
index ba77adf..58401bc 100644
--- a/src/virtManager/console.py
+++ b/src/virtManager/console.py
@@ -1194,6 +1194,8 @@ class vmmConsolePages(vmmGObjectUI):
         if self.widget("console-auth-remember").get_active():
             self.config.set_console_password(self.vm, passwd.get_text(),
                                              username.get_text())
+        else:
+            self.config.del_console_password(self.vm)
 
     def queue_scroll_resize_helper(self, w, h):
         """
diff --git a/src/virtManager/keyring.py b/src/virtManager/keyring.py
index 43d1574..281ac92 100644
--- a/src/virtManager/keyring.py
+++ b/src/virtManager/keyring.py
@@ -67,6 +67,12 @@ class vmmKeyring(object):
 
         return _id
 
+    def del_secret(self, _id):
+        try:
+            gnomekeyring.item_delete_sync(self.keyring, _id)
+        except:
+            logging.exception("Failed to delete keyring secret")
+
     def get_secret(self, _id):
         sec = None
         try:
diff --git a/src/vmm-details.glade b/src/vmm-details.glade
index 2f78b39..8843422 100644
--- a/src/vmm-details.glade
+++ b/src/vmm-details.glade
@@ -614,6 +614,7 @@
                         <property name="label" translatable="yes">_Save this password in your keyring</property>
                         <property name="visible">True</property>
                         <property name="can_focus">True</property>
+                        <property name="tooltip_text" translatable="yes">Check to save password, uncheck to forget password.</property>
                         <property name="receives_default">False</property>
                         <property name="use_underline">True</property>
                         <property name="yalign">0</property>
-- 
2.10.0

